discord/event.py

```diff
@@ -0,0 +1,273 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import datetime
+    from typing import Optional, Sequence
+
+    from discord import Emoji, Guild, GuildSticker, Integration, Invite, RawIntegrationDeleteEvent, User
+    from discord.abc import GroupChannel, GuildChannel
+
+__all__ = (
+    'Event',
+    'GuildChannelDelete',
+    'GuildChannelCreate',
+    'GuildChannelUpdate',
+    'GroupJoin',
+    'GroupRemove',
+    'GuildChannelPinsUpdate',
+    'PrivateChannelUpdate',
+    'PrivateChannelPinsUpdate',
+    'Typing',
+    'Connect',
+    'Disconnect',
+    'ShardConnect',
+    'ShardDisconnect',
+    'Error',
+    'SocketEventType',
+    'SocketRawReceive',
+    'SocketRawSend',
+    'Ready',
+    'Resumed',
+    'ShardReady',
+    'ShardResumed',
+    'GuildAvailable',
+    'GuildUnavailable',
+    'GuildJoin',
+    'GuildRemove',
+    'GuildUpdate',
+    'GuildEmojisUpdate',
+    'GuildStickersUpdate',
+    'InviteCreate',
+    'InviteDelete',
+    'IntegrationCreate',
+    'IntegrationUpdate',
+    'GuildIntegrationUpdate',
+    'WekhooksUpdate',
+    'RawIntegrationDelete',
+    'Interaction',
+)
+
+
+class Event(ABC):
+    pass
+
+
+@dataclass(frozen=True)
+class GuildChannelDelete(Event):
+    channel: GuildChannel
+
+
+@dataclass(frozen=True)
+class GuildChannelCreate(Event):
+    channel: GuildChannel
+
+
+@dataclass(frozen=True)
+class GuildChannelUpdate(Event):
+    before: GuildChannel
+    after: GuildChannel
+
+
+@dataclass(frozen=True)
+class GroupJoin(Event):
+    channel: GroupChannel
+    user: User
+
+
+@dataclass(frozen=True)
+class GroupRemove(Event):
+    channel: GroupChannel
+    user: User
+
+
+@dataclass(frozen=True)
+class GuildChannelPinsUpdate(Event):
+    channel: GuildChannel
+    last_pin: Optional[datetime.datetime]
+
+
+@dataclass(frozen=True)
+class PrivateChannelUpdate(Event):
+    before: GroupChannel
+    after: GroupChannel
+
+
+@dataclass(frozen=True)
+class PrivateChannelPinsUpdate(Event):
+    channel: GroupChannel
+    last_pin: Optional[datetime.datetime]
+
+
+@dataclass(frozen=True)
+class Typing(Event):
+    pass
+
+
+@dataclass(frozen=True)
+class Connect(Event):
+    pass
+
+
+@dataclass(frozen=True)
+class Disconnect(Event):
+    pass
+
+
+@dataclass(frozen=True)
+class ShardConnect(Event):
+    shard_id: int
+
+
+@dataclass(frozen=True)
+class ShardDisconnect(Event):
+    shard_id: int
+
+
+@dataclass(frozen=True)
+class Error(Event):
+    event: Event
+
+
+@dataclass(frozen=True)
+class SocketEventType(Event):
+    event_type: str
+
+
+@dataclass(frozen=True)
+class SocketRawReceive(Event):
+    msg: str
+
+
+@dataclass(frozen=True)
+class SocketRawSend(Event):
+    payload: str | bytes

```

Comment:
This is a Python 3.10+ syntax, while discord.py 2.0 targets Python 3.9

Reply:
3.8, actually! :)

Reply:
Oh I did it again, sorry for that!

Reply:
By using ``from __future__ import annotations`` this syntax is actually valid, however it's inconsistent with the rest of the library and should be replaced with a ``Union[str, bytes]`` regardless.